### PR TITLE
Use a dynamic notion of distance

### PR DESCRIPTION
**Pull request by ltratt**
This PR consists of two initial refactoring commits (both very easy) before the "big one" (though it's not big in terms of code): moving to a dynamic notion of distance in the A* algorithm. The commit message for https://github.com/softdevteam/lrpar/commit/b42e1b4c0b3812173b775f22c630181ddb9c0d94 hopefully explains this in sufficient detail.

### CODE REVIEW

**Comment by ptersilie** [comment]
I'm not sure I can follow the explanation of the last commit. If `U` follows `T` how can the distance to `U` (1) be smaller than the distance to `T` (2)? Shouldn't the distance to `U` be the distance to `T` plus the distance to `U`?

**Comment by ltratt** [comment]
Good question. I didn't explain one important detail.

If the input is valid, then U's distance must always be > than T's. *But* if the input is invalid, the user might have put a silly T in that has a greater distance than U.

Imagine if I had a Java input something like `var x = } 5;`. I think I'm right in saying that the distance from `=` to `}` is 2, but the distance from `=` to `5` is 0.

File: src/lib/mf.rs [hunk 1]
```diff
@@ -457,6 +463,73 @@ impl<'a, TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + Partial
...
+                all_shfts = false;
+                break;
+            }
+            if all_shfts {
+                return Some(0);
+            }
+        }
+
+        // Now we deal with the "main" case: dealing with distances in the face of possible
+        // deletions. Imagine we there are two lexemes starting at position la_idx: (in order) T
```
Comment (ptersilie) [inline src/lib/mf.rs]: ~~we~~

**Comment by ptersilie** [comment]
Aaaaahhhhh! That explains everything. :)

**Comment by ltratt** [comment]
If it's any consolation, I first had this rough thought over 4 weeks ago, but it took me ages to work out how to trigger this in practise... It's really not obvious (at least... it wasn't to me!).

File: src/lib/mf.rs [hunk 2]
```diff
@@ -457,6 +463,73 @@ impl<'a, TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + Partial
...
+        // dist() from st_idx to U is 0. If we delete T then the distance to U is 1, which is a
+        // shorter distance than T. We therefore need to return a distance of 1, even though that
+        // is the distance to the second lexeme.
+        let frst_tidx = self.parser.next_lexeme(None, la_idx).1;
+        let mut ld = self.dist.dist(st_idx, frst_tidx)
+                              .unwrap_or(u32::max_value()); // ld == Current least distance
+        let mut dc = (self.parser.term_cost)(frst_tidx) as u32; // Cumulative deletion cost
+        for i in la_idx + 1..self.parser.lexemes.len() + 1 {
+            if dc >= ld {
+                // One the cumulative cost of deleting lexemes is bigger than the current least
```
Comment (ptersilie) [inline src/lib/mf.rs]: ~~One~~ +When

**Comment by ptersilie** [comment]
It's obvious when you explain it like this. But I can see how finding this example isn't obvious at all.

**Comment by ptersilie** [comment]
So just a few typos. Rest looks fine.

**Comment by ltratt** [comment]
OK, fixed both typos. Can squash if you'd like?

**Comment by ltratt** [comment]
Actually, maybe I should add that Java example in to the comment first...

**Comment by ltratt** [comment]
OK, example added. If you think that's good, I can squash.

**Comment by ptersilie** [comment]
Yes, I think the java example is really helpful for the understanding in that commit. I'm happy with this. Squash away and I will merge.

**Comment by ltratt** [comment]
Squashed.

**Comment by ptersilie** [comment]
Nice. Merged.

**Pull request merged**